aws/rust-runtime/aws-inlineable/src/s3_express.rs

```diff
@@ -103,23 +103,383 @@ pub(crate) mod auth {
 
 /// Supporting code for S3 Express identity cache
 pub(crate) mod identity_cache {
+    use aws_credential_types::Credentials;
+    use aws_smithy_async::time::SharedTimeSource;
+    use aws_smithy_runtime::expiring_cache::ExpiringCache;
+    use aws_smithy_runtime_api::box_error::BoxError;
+    use aws_smithy_runtime_api::client::identity::Identity;
+    use hmac::{digest::FixedOutput, Hmac, Mac};
+    use lru::LruCache;
+    use sha2::Sha256;
+    use std::fmt;
+    use std::future::Future;
+    use std::hash::Hash;
+    use std::sync::Mutex;
+    use std::time::{Duration, SystemTime};
+
+    pub(crate) const DEFAULT_MAX_CACHE_CAPACITY: usize = 100;
+    pub(crate) const DEFAULT_BUFFER_TIME: Duration = Duration::from_secs(10);
+
+    #[derive(Clone, Eq, PartialEq, Hash)]
+    pub(crate) struct CacheKey(String);
+
+    impl CacheKey {
+        pub(crate) fn from(bucket_name: &str, creds: &Credentials) -> Self {
+            Self({
+                let key = format!("{}{}", creds.access_key_id(), creds.secret_access_key());
+                let mac = Hmac::<Sha256>::new_from_slice(key.as_ref())
+                    .expect("HMAC can take key of any size");
+                let mut inner = hex::encode(mac.finalize_fixed());
+                inner.push_str(bucket_name);
+                inner
+            })
+        }
+    }
+
     /// The caching implementation for S3 Express identity.
     ///
     /// While customers can either disable S3 Express itself or provide a custom S3 Express identity
     /// provider, configuring S3 Express identity cache is not supported. Thus, this is _the_
     /// implementation of S3 Express identity cache.
-    #[derive(Debug)]
-    pub(crate) struct S3ExpressIdentityCache;
-}
+    pub(crate) struct S3ExpressIdentityCache {
+        pub(crate) inner: Mutex<LruCache<CacheKey, ExpiringCache<Identity, BoxError>>>,
+        pub(crate) time_source: SharedTimeSource,
+        pub(crate) buffer_time: Duration,
+    }
+
+    impl fmt::Debug for S3ExpressIdentityCache {
+        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+            f.debug_struct("S3ExpressIdentityCache")
+                .field("time_source", &self.time_source)
+                .field("buffer_time", &self.buffer_time)
+                .finish()
+        }
+    }
+
+    impl S3ExpressIdentityCache {
```

Comment:
Nit: Add a `new` function to this so that construction is easier in DefaultS3ExpiringIdentityProvider.

Reply:
Added in 86e33cc.